File: apps/backend-java/src/main/java/com/housingplatform/features/payments/PaymentService.java
package com.housingplatform.features.payments;

import com.housingplatform.shared.auth.error.BadRequestException;
import com.housingplatform.shared.auth.error.ExternalServiceException;
import com.housingplatform.shared.auth.error.ForbiddenException;
import com.housingplatform.shared.auth.error.NotFoundException;
import com.housingplatform.shared.auth.error.PaymentAmountMismatchException;
import com.housingplatform.shared.auth.error.PaymentFailedException;
import com.housingplatform.shared.auth.model.AuthenticatedUser;
import com.housingplatform.persistence.repository.PaymentRepository;
import com.housingplatform.features.payments.dto.ConfirmPaymentRequest;
import com.housingplatform.features.payments.dto.ConfirmPaymentResult;
import com.housingplatform.features.payments.dto.CreatePaymentOrderRequest;
import com.housingplatform.features.payments.dto.CreatePaymentOrderResult;
import com.housingplatform.features.payments.dto.TossWebhookPayload;
import com.housingplatform.features.payments.dto.WebhookAck;
import com.housingplatform.features.payments.dto.WebhookAckStatus;
import com.housingplatform.features.payments.mapper.PaymentMapper;
import com.housingplatform.persistence.enums.PaymentStatus;
import com.housingplatform.shared.RateLimitService;
import java.util.Map;
import java.util.UUID;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class PaymentService {

  private final PaymentRepository paymentRepository;
  private final PaymentFinalizationService paymentFinalizationService;
  private final RateLimitService rateLimitService;
  private final TossClient tossClient;

  public PaymentService(
      PaymentRepository paymentRepository,
      PaymentFinalizationService paymentFinalizationService,
      RateLimitService rateLimitService,
      TossClient tossClient) {
    this.paymentRepository = paymentRepository;
    this.paymentFinalizationService = paymentFinalizationService;
    this.rateLimitService = rateLimitService;
    this.tossClient = tossClient;
  }

  @Transactional
  public CreatePaymentOrderResult createPaymentOrder(
      AuthenticatedUser user, CreatePaymentOrderRequest request) {
    rateLimitService.assertRateLimit("create-payment:" + user.id(), 20, 60);
    var order = paymentRepository.createPaymentOrder(request.bookingId(), user.id());
    return PaymentMapper.toCreateOrderResult(
        order.paymentId(), order.orderId(), order.bookingId(), order.amountKrw(), order.orderName());
  }

  @Transactional
  public ConfirmPaymentResult confirmPayment(AuthenticatedUser user, ConfirmPaymentRequest request) {
    rateLimitService.assertRateLimit("confirm-payment:" + user.id(), 20, 60);

    var payment =
        paymentRepository
            .findByOrderId(request.orderId())
            .orElseThrow(() -> new NotFoundException("Payment not found"));

    if (!payment.customerId().equals(user.id())) {
      throw new ForbiddenException("You cannot confirm this payment");
    }
    if (payment.amountKrw() != request.amount()) {
      throw new PaymentAmountMismatchException("Payment amount does not match booking total");
    }
    if (payment.status() == PaymentStatus.confirmed) {
      return PaymentMapper.toConfirmResult(
          payment.id(), payment.orderId(), null, PaymentStatus.confirmed);
    }

    Map<String, Object> tossResponse = null;
    try {
      tossResponse =
          tossClient.confirmPayment(
              request.paymentKey(), request.orderId().toString(), request.amount());
    } catch (TossClient.TossClientError exception) {
      paymentFinalizationService.markPaymentFailed(
          request.orderId(), exception.getMessage(), tossResponse);
      throw new PaymentFailedException(exception.getMessage());
    }

    if (!TossClient.isSuccessful(tossResponse)) {
      String reason = String.valueOf(tossResponse.getOrDefault("status", "Payment not completed"));
      paymentFinalizationService.markPaymentFailed(request.orderId(), reason, tossResponse);
      throw new PaymentFailedException("Payment was not completed");
    }

    var finalized =
        paymentFinalizationService.finalizeSuccessfulPayment(
            request.orderId(), request.paymentKey(), request.amount(), tossResponse);
    return PaymentMapper.toConfirmResult(finalized);
  }

  @Transactional
  public WebhookAck receiveWebhook(TossWebhookPayload payload) {
    if (tossClient.isDevMockEnabled() && !tossClient.hasSecretKey()) {
      return new WebhookAck(true, WebhookAckStatus.ignored);
    }

    Map<String, Object> data = payload.data() == null ? Map.of() : payload.data();
    String paymentKey = String.valueOf(data.getOrDefault("paymentKey", ""));
    String orderIdRaw = String.valueOf(data.getOrDefault("orderId", ""));

    if (paymentKey.isBlank() || orderIdRaw.isBlank()) {
      throw new BadRequestException("Webhook payload missing paymentKey or orderId");
    }

    UUID orderId;
    try {
      orderId = UUID.fromString(orderIdRaw);
    } catch (IllegalArgumentException exception) {
      throw new BadRequestException("Webhook payload missing paymentKey or orderId");
    }

    String eventId =
        (payload.eventType() == null ? "UNKNOWN" : payload.eventType())
            + ":"
            + paymentKey
            + ":"
            + (payload.createdAt() == null ? "unknown" : payload.createdAt());

    var payment =
        paymentRepository
            .findByOrderId(orderId)
            .orElseThrow(() -> new NotFoundException("Payment not found"));

    paymentFinalizationService.recordPaymentEvent(
        eventId,
        payment.id(),
        payment.bookingId(),
        payload.eventType() == null ? "UNKNOWN" : payload.eventType(),
        Map.of(
            "eventType", payload.eventType(),
            "createdAt", payload.createdAt(),
            "data", data));

    if (payment.status() == PaymentStatus.confirmed) {
      return new WebhookAck(true, WebhookAckStatus.already_confirmed);
    }

    Map<String, Object> tossPayment;
    try {
      tossPayment = tossClient.fetchPayment(paymentKey);
    } catch (TossClient.TossClientError exception) {
      throw new ExternalServiceException(exception.getMessage());
    }

    if (TossClient.isSuccessful(tossPayment)) {
      int amount = ((Number) tossPayment.getOrDefault("totalAmount", payment.amountKrw())).intValue();
      paymentFinalizationService.finalizeSuccessfulPayment(
          orderId, paymentKey, amount, tossPayment);
      return new WebhookAck(true, WebhookAckStatus.confirmed);
    }

    if (TossClient.isFailed(tossPayment)) {
      paymentFinalizationService.markPaymentFailed(
          orderId, String.valueOf(tossPayment.getOrDefault("status", "Payment failed")), tossPayment);
      return new WebhookAck(true, WebhookAckStatus.failed);
    }

    return new WebhookAck(true, WebhookAckStatus.ignored);
  }
}

File: apps/backend-node/src/features/payments/payment.service.ts
import { Injectable } from '@nestjs/common';

import type { AuthUser } from '@/shared/auth/auth-user.model';
import {
  BadRequestError,
  ExternalServiceError,
  ForbiddenError,
  NotFoundError,
  PaymentAmountMismatchError,
  PaymentFailedError,
} from '@/shared/errors';
import { RateLimitService } from '@/shared/rate-limit/rate-limit.service';

import type {
  ConfirmPaymentRequest,
  ConfirmPaymentResult,
  CreatePaymentOrderRequest,
  CreatePaymentOrderResult,
  TossWebhookPayload,
  WebhookAck,
} from './dto';
import { PaymentStatus, WebhookAckStatus } from './dto';
import {
  mapConfirmPaymentResult,
  mapCreatePaymentOrderResult,
} from './mappers/payment.mapper';
import { PaymentFinalizationService } from './payment-finalization.service';
import { PaymentRepository } from './payment.repository';
import { TossClient, TossClientError } from './toss.client';

@Injectable()
export class PaymentService {
  constructor(
    private readonly repository: PaymentRepository,
    private readonly tossClient: TossClient,
    private readonly finalization: PaymentFinalizationService,
    private readonly rateLimit: RateLimitService,
  ) {}

  async createPaymentOrder(
    user: AuthUser,
    request: CreatePaymentOrderRequest,
  ): Promise<CreatePaymentOrderResult> {
    await this.rateLimit.assertRateLimit(`create-payment:${user.id}`, 20, 60);

    const order = await this.repository.createPaymentOrder(
      request.bookingId,
      user.id,
    );

    return mapCreatePaymentOrderResult({
      paymentId: order.paymentId,
      orderId: order.orderId,
      bookingId: order.bookingId,
      amountKrw: order.amountKrw,
      orderName: order.orderName,
    });
  }

  async confirmPayment(
    user: AuthUser,
    request: ConfirmPaymentRequest,
  ): Promise<ConfirmPaymentResult> {
    await this.rateLimit.assertRateLimit(`confirm-payment:${user.id}`, 20, 60);

    const payment = await this.repository.getPaymentByOrderId(request.orderId);
    if (!payment) {
      throw new NotFoundError('Payment not found');
    }

    if (payment.customerId !== user.id) {
      throw new ForbiddenError('You cannot confirm this payment');
    }

    if (payment.amountKrw !== request.amount) {
      throw new PaymentAmountMismatchError();
    }

    if (payment.status === 'confirmed') {
      return {
        paymentId: payment.id,
        orderId: payment.orderId,
        bookingId: null,
        status: PaymentStatus.Confirmed,
      };
    }

    let tossResponse: Record<string, unknown> | null = null;

    try {
      tossResponse = await this.tossClient.confirmPayment({
        paymentKey: request.paymentKey,
        orderId: request.orderId,
        amount: request.amount,
      });
    } catch (error) {
      if (error instanceof TossClientError) {
        await this.finalization.markPaymentFailed({
          orderId: request.orderId,
          reason: error.message,
          tossResponse,
        });
        throw new PaymentFailedError(error.message);
      }
      throw error;
    }

    if (!TossClient.isSuccessful(tossResponse)) {
      const reason = String(tossResponse.status ?? 'Payment not completed');
      await this.finalization.markPaymentFailed({
        orderId: request.orderId,
        reason,
        tossResponse,
      });
      throw new PaymentFailedError('Payment was not completed');
    }

    const finalized = await this.finalization.finalizeSuccessfulPayment({
      orderId: request.orderId,
      paymentKey: request.paymentKey,
      amountKrw: request.amount,
      tossResponse,
    });

    return mapConfirmPaymentResult(finalized);
  }

  async receiveWebhook(payload: TossWebhookPayload): Promise<WebhookAck> {
    if (this.tossClient.isDevMockEnabled() && !this.tossClient.hasSecretKey()) {
      return { ok: true, status: WebhookAckStatus.Ignored };
    }

    const data = payload.data ?? {};
    const paymentKey = String(data.paymentKey ?? '');
    const orderIdRaw = String(data.orderId ?? '');

    if (!paymentKey || !orderIdRaw) {
      throw new BadRequestError('Webhook payload missing paymentKey or orderId');
    }

    if (!/^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i.test(orderIdRaw)) {
      throw new BadRequestError('Webhook payload missing paymentKey or orderId');
    }

    const eventId = `${payload.eventType ?? 'UNKNOWN'}:${paymentKey}:${payload.createdAt ?? 'unknown'}`;

    const payment = await this.repository.getPaymentByOrderId(orderIdRaw);
    if (!payment) {
      throw new NotFoundError('Payment not found');
    }

    await this.finalization.recordPaymentEvent({
      eventId,
      paymentId: payment.id,
      bookingId: payment.bookingId,
      eventType: payload.eventType ?? 'UNKNOWN',
      payload: payload as Record<string, unknown>,
    });

    if (payment.status === 'confirmed') {
      return { ok: true, status: WebhookAckStatus.AlreadyConfirmed };
    }

    let tossPayment: Record<string, unknown>;
    try {
      tossPayment = await this.tossClient.fetchPayment(paymentKey);
    } catch (error) {
      if (error instanceof TossClientError) {
        throw new ExternalServiceError(error.message);
      }
      throw error;
    }

    if (TossClient.isSuccessful(tossPayment)) {
      const amount = Number(tossPayment.totalAmount ?? payment.amountKrw);
      await this.finalization.finalizeSuccessfulPayment({
        orderId: orderIdRaw,
        paymentKey,
        amountKrw: amount,
        tossResponse: tossPayment,
      });
      return { ok: true, status: WebhookAckStatus.Confirmed };
    }

    if (TossClient.isFailed(tossPayment)) {
      await this.finalization.markPaymentFailed({
        orderId: orderIdRaw,
        reason: String(tossPayment.status ?? 'Payment failed'),
        tossResponse: tossPayment,
      });
      return { ok: true, status: WebhookAckStatus.Failed };
    }

    return { ok: true, status: WebhookAckStatus.Ignored };
  }
}
